### PR TITLE
Implement ctrl+scroll and pinch to zoom gestures

### PR DESCRIPTION
**Pull request by Sjoerd1993**
See title. 

Still need to change the zooming logic so that it works on non-linear axes. Also, the way to detect if a ctrl is pressed is not my ideal, but for some reason the key_press event in the canvas is not registered. My guess is that this is because the canvas is not in focus, which is by design as we need to catch the shortkeys in the main application.

[Screencast from 2023-09-18 14-39-55.webm](https://github.com/Sjoerd1993/Graphs/assets/68477016/58bf3b36-5f41-4b9f-b512-1474a63d6f54)

Closes issue #491, also closes the last checkbox in #385 but I won't close that one until we've been accepted into GNOME circle, as new feedback will probably arise.

### CODE REVIEW

**Comment by Sjoerd1993** [comment]
This works now, I'm open to changing ctrl+scroll to just scrolling without having to hold control by the way.

**Comment by cmkohnen** [comment]
As per [GNOME HIG](https://developer.gnome.org/hig/guidelines/pointer-touch.html#scrolling-panning-zooming) Ctrl + Scroll to zoom is intended for views, that can scroll vertically (e. g. viewing a pdf document) and for views that won't scroll vertically (like in our case), just using the scroll wheel should suffice.

File: src/canvas.py [hunk 1]
```diff
@@ -104,11 +106,99 @@ def __init__(self, application):
...
+        scale = 1.6
+        if event.button == "up":
+            scale = 1 / scale
+        if self.get_application().get_ctrl() is True:
+            self.on_zoom_event(scale)
+
+    def on_zoom_event(self, scaling=1.6):
+        for ax in [self.axis]:
+            zoom_factor = scaling
+            scale_mapping = {
```
Comment (cmkohnen) [inline src/canvas.py:151]: I'd use scales.to_int() here.

File: src/canvas.py [hunk 2]
```diff
@@ -104,11 +106,99 @@ def __init__(self, application):
...
+
+    def on_scroll_event(self, event):
+        scale = 1.6
+        if event.button == "up":
+            scale = 1 / scale
+        if self.get_application().get_ctrl() is True:
+            self.on_zoom_event(scale)
+
+    def on_zoom_event(self, scaling=1.6):
+        for ax in [self.axis]:
```
Comment (cmkohnen) [inline src/canvas.py:149]: shouldn't this be applied to all axes?

Comment (Sjoerd1993) [reply src/canvas.py:149]: Thanks, yes I put this to only the first axis for testing purposes. Forgot to bring it back

File: src/canvas.py (hunk 1 above)
Comment (Sjoerd1993) [reply src/canvas.py:151]: Thanks, just didn't think about this function, that's obviously a much better choice

File: src/canvas.py [hunk 3]
```diff
@@ -104,11 +106,99 @@ def __init__(self, application):
...
+        self.add_controller(zoom_gesture)
 
     def get_application(self):
         """Get application property."""
         return self.props.application
 
+    def get_coord_fraction(self, event):
+        xlim = self.top_right_axis.get_xlim()
+        ylim = self.top_right_axis.get_ylim()
+        figure_settings = \
```
[the remark this answers is not in the document]
Comment (Sjoerd1993) [reply src/canvas.py:124]: Also here, yes that's cleaner obviously.

**Comment by Sjoerd1993** [comment]
I also added ctrl+ and ctrl- for zoom in and zoom out respectively. The ctrl+ and ctrl- action zoom into the middle, not where the mouse pointer is pointing. This is also what Inkscape does for instance. But I'm open to let it zoom to where the mouse pointer is pointing, and go for the middle if the mouse pointer is not on the canvas.

The used triggers to match to the different plus/minus keys were taken from [Nautilus](https://github.com/GNOME/nautilus/blob/ca829d48a01df9c9db9e5f68cccb2244874917b5/src/nautilus-files-view.c#L9949), hence the "ZoomIn" and "ZoomOut" accelerators in there as well.

**Review by cmkohnen — APPROVED** [review]
Both scolling and Pinch to zoom work in my testing. However two finger scrolling is quite aggressive. Don't know if that can be fixed however.

The keyboard shortcuts behaviour is fine as is, using the mouse position there is unintuitive

**Comment by Sjoerd1993** [comment]
> Both scolling and Pinch to zoom work in my testing. However two finger scrolling is quite aggressive. Don't know if that can be fixed however.

Yeah, it's why I reduced the zoom factor for the scroll event quite a bit compared to the original PR. I tuned them down a bit more now as well.

**Pull request merged**